### PR TITLE
[JENKINS-31545] Added JiraIssueUpdater publisher

### PR DESCRIPTION
**Pull request by v1v**
Added [JiraIssueUpdater publisher](https://wiki.jenkins-ci.org/display/JENKINS/JIRA+Plugin) plugin support.

- https://issues.jenkins-ci.org/browse/JENKINS-31545

In order to use this feature you need to configure your Jenkins global settings:
- https://wiki.jenkins-ci.org/display/JENKINS/JIRA+Plugin#JIRAPlugin-UpdatingJIRAissueswithbackpointers

### CODE REVIEW

**Comment by v1v** [comment]
I've been debugging the jira plugin and I've found the below jobs configuration:

Builder step:
```xml
<builders>
<hudson.plugins.jira.JiraIssueUpdateBuilder plugin="jira@1.39">
<jqlSearch>bla</jqlSearch>
<workflowActionName>bla</workflowActionName>
<comment>bla</comment>
</hudson.plugins.jira.JiraIssueUpdateBuilder>
</builders>
```

Publisher section:
```xml
<publishers>
<hudson.plugins.jira.JiraIssueUpdater plugin="jira@1.39"/>
<hudson.plugins.jira.JiraVersionCreator plugin="jira@1.39">
<jiraVersion>bla</jiraVersion>
<jiraProjectKey>bla</jiraProjectKey>
</hudson.plugins.jira.JiraVersionCreator>
<hudson.plugins.jira.JiraCreateIssueNotifier plugin="jira@1.39">
<projectKey>bla</projectKey>
<testDescription>bla</testDescription>
<assignee>bla</assignee>
<component>bla</component>
</hudson.plugins.jira.JiraCreateIssueNotifier>
<hudson.plugins.jira.JiraReleaseVersionUpdater plugin="jira@1.39">
<jiraProjectKey>bla</jiraProjectKey>
<jiraRelease>bla</jiraRelease>
</hudson.plugins.jira.JiraReleaseVersionUpdater>
<hudson.plugins.jira.JiraIssueMigrator plugin="jira@1.39">
<jiraProjectKey>bla</jiraProjectKey>
<jiraRelease>bla</jiraRelease>
<jiraReplaceVersion>bla</jiraReplaceVersion>
<jiraQuery>bla</jiraQuery>
</hudson.plugins.jira.JiraIssueMigrator>
</publishers>
````

Wrapper section:
```xml
<buildWrappers>
<hudson.plugins.jira.JiraCreateReleaseNotes plugin="jira@1.39">
<jiraEnvironmentVariable>bla</jiraEnvironmentVariable>
<jiraProjectKey>bla</jiraProjectKey>
<jiraRelease>bla</jiraRelease>
<jiraFilter>bla</jiraFilter>
</hudson.plugins.jira.JiraCreateReleaseNotes>
</buildWrappers>
```


Do you think we should use the below DSL clousure?

```
job('example') {
 
    wrapper {
       jira (...) {

       }
    }
    steps {
       jira (...) {

       }
    }
    publishers {
       jira (...) {

       }
    }
}
```

Let me know your thoughts/concerns

cheers

**Comment by v1v** [comment]
I reworked on this PR and now it does fully support the Jira plugin (version 1.39 onwards) based on the above dsl syntax.

Cheers

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/step/StepContext/jira.groovy [hunk 1]
```diff
@@ -0,0 +1,11 @@
+job('example') {
+    steps {
+        jira {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/step/StepContext/jira.groovy]: do not add a `jira` context here

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/step/StepContext/jira.groovy [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+            jiraIssueUpdateBuilder {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/step/StepContext/jira.groovy]: rename this to `progressJiraIssues`

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext/jira.groovy [hunk 3]
```diff
@@ -0,0 +1,12 @@
+job('example') {
+    wrappers {
+        jira {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext/jira.groovy]: do not add a `jira` context here

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext/jira.groovy [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+            jiraCreateReleaseNotes {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext/jira.groovy]: rename this to `generateJiraReleaseNotes`

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext/jira.groovy [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+                jiraEnvironmentVariable('bla')
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext/jira.groovy]: remove the `jira` prefix for all methods in this context

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 6]
```diff
@@ -0,0 +1,27 @@
+job('example') {
+    publishers {
+        jira {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: do not add a `jira` context here

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+            jiraIssueUpdater()
+            jiraReleaseVersionUpdater {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: rename this to `releaseJiraVersion`

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+                jiraProjectKey('bla')
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: remove this `jira` prefix for all methods in this context

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+                jiraRelease('bla')
+            }
+            jiraIssueMigrator {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: rename this to `moveJiraIssues`

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+                jiraProjectKey('bla')
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: remove the `jira` prefix for all methods in this context

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+                jiraRelease('bla')
+                jiraReplaceVersion('bla')
+                jiraQuery('bla')
+            }
+            jiraCreateIssueNotifier {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: rename this to `createJiraIssue`

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+                projectKey('bla')
+                testDescription('bla')
+                assignee('bla')
+                component('bla')
+            }
+            jiraVersionCreator {
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: rename this to `createJiraVersion`

File: job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+                jiraVersion('bla')
```
Comment (daspilker) [inline job-dsl-core/src/main/docs/examples/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext/jira.groovy]: remove the `jira` prefix for all methods in this context

**Comment by daspilker** [comment]
Do you want to pick this up? Otherwise I will close the PR.

**Comment by v1v** [comment]
Sorry about my delay, I'll be working on it shortly

**Comment by v1v** [comment]
Already replaced those Jira closures, renamed and rebased from the master branch. I've also verified it works in a test environment.

Cheers

**Pull request merged**